src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpAdapter.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Adapter\PhpAmqp;
+
+use Symfony\Component\Messenger\Adapter\Factory\AdapterInterface;
+use Symfony\Component\Messenger\Transport\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\SenderInterface;
+use Symfony\Component\Messenger\Transport\Serialization\DecoderInterface;
+use Symfony\Component\Messenger\Transport\Serialization\EncoderInterface;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class AmqpAdapter implements AdapterInterface
+{
+    private $encoder;
+    private $decoder;
+    private $connection;
+
+    public function __construct(EncoderInterface $encoder, DecoderInterface $decoder, Connection $connection)
+    {
+        $this->encoder = $encoder;
+        $this->decoder = $decoder;
+        $this->connection = $connection;
+    }
+
+    public function receiver(): ReceiverInterface
+    {
+        return new AmqpReceiver($this->decoder, $this->connection);
```

Comment:
This is actually registered as a service, and therefore executed once by the DIC :)

Reply:
The `AmqpAdapter` is registered in the DIC, not the `AmqpReceiver` isn't ?

Reply:
It is: check [this](https://github.com/symfony/symfony/pull/26632/files#diff-0e793081ceb720201745c982a568903fR1470) out. The adapter service is the factory of the receiver and sender services.

Reply:
Sorry, I guess I missed something :( The `AmqpAdapter`, the sender and the receiver are indeed registred in the DIC. This **this** very `AmqpAdapter` does not use the the sender or the receiver that are in the DIC.

Reply:
This very `AmqpAdapter` is used to create the sender and receiver services. Therefore `$container->get('messenger.[name]_sender')` will get this very `AmqpSender` object.

Though, you are right that `$this->get('messenger.[name]_adapter')->receiver()` will return a new instance, and this might be a cause of issues. I guess I'll add a `if (null === $this->receiver) ...` in the adapter so it won't re-create it.